README
NAME
    Dist::Zilla::App::Command::self - Build a distribution with a
    bootstrapped version of itself.

VERSION
    version 0.001003

SYNOPSIS
    This is a different approach to using "[Bootstrap::lib]" that absolves a
    distribution from needing to forcibly embed bootstrapping logic in
    "dist.ini"

      dzil self build

    This is largely similar to using

      [Bootstrap::lib]
      try_built = 1

    and doing

      dzil build && dzil build

    And similar again to:

      dzil run bash -c "cd ../; dzil -I$BUILDDIR/lib dzil build"

    Or whatever the magic is that @ETHER uses.

    This also means that:

      dzil self release

    Is something you can do.

CAVEATS
    The nature of this implies that your distribution will probably need an
    older generation of itself for the initial bootstrap.

    That is to say:

      dzil build

    Must work, and use "Generation.Previous" to build "Generation.Build"

      dzil self foo

    Will call "dzil build" for you, to build "Generation.Build", and then
    invoke

      dzil foo

    To use "Generation.Build" to build "Generation.Next"

    "1. Generation.Previous"
        A previously installed incarnation of your dist.

    "2. Generation.Build"
        The iteration of building the distribution itself from source using
        "Generation.Previous"

    "3. Generation.Next"
        The iteration of building the distribution itself from source using
        "Generation.Build"

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Dist-Zilla-App-Command-self

0.001004 2017-03-07T17:09:40Z af19241

0.001003 2017-03-07T17:08:39Z e44adf0
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - develop: +8 ↑4 -2 (recommends: ↓1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: ↓1 (recommends: ↓1)

0.001002 2014-10-31T09:37:01Z bbef681
 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +5 ↑1 -1 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1
 - test: +1 (recommends: ↑2)

 [Internals]
 - remove utf8 requirement
 - now works on 5.006

 [Metadata]
 - author = kentnl

0.001001 2014-08-16T23:36:01Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests updated.
 - metadata updated.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +5 ↑1 -3 (suggests: +2 -1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - move to kentnl org.

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

0.001000 2014-05-22T14:21:45Z
 - First version.


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw());
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Build a distribution with a bootstrapped version of itself.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-App-Command-self",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::App::Command::self",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dist::Zilla::App" => 0,
    "File::pushd" => 0,
    "Path::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dist::Zilla::App" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::pushd" => 0,
  "Path::Tiny" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
